File: Blind75/297. Serialize and Deserialize Binary Tree/297. Serialize and Deserialize Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    string serialize(TreeNode* root){
        if(root==NULL) return "x,";
        // preorder traverse
        return to_string(root->val)+ "," + serialize(root->left) + serialize(root->right);
    }
    queue<string> q;
    TreeNode* deserialize(string data) {
        int n = data.size();
        cout<<data<<endl;
        for(int i=0; i<n; i++){
            //find next ','
            int r = i+1;
            while(r<n && data[r]!=',') r++;
            q.push(data.substr(i, r-i));
            i = r;
        }

        return tree(q);

    }
    
    TreeNode* tree(queue<string>& q){
        //base case
        if(q.empty()) return NULL;
        if(q.front()=="x"){
            q.pop();
            return NULL;
        }

        TreeNode* node = new TreeNode(stoi(q.front()));
        q.pop();
        node->left = tree(q);
        node->right = tree(q);

        return node;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: Blind75/297. Serialize and Deserialize Binary Tree/297. Serialize and Deserialize Binary Tree.cs
public class Codec {

    // Encodes a tree to a single string.
    public string serialize(TreeNode root) {
        if(root==null) return "x,";
        return root.val.ToString() + "," + serialize(root.left)+serialize(root.right);
    }
    Queue<string> Q = new Queue<string>();
    // Decodes your encoded data to tree.
    public TreeNode deserialize(string data) {
        int n = data.Length;
        for(int i=0; i<n; i++){
            int r = i+1;
            while(r<n && data[r]!=',') r++;
            string tmp = data.Substring(i,r-i);
            Q.Enqueue(tmp);
            i=r;
        }
        
        

        //return null;
        return tree(ref Q);
    }


    public TreeNode tree(ref Queue<string> Q){
        //if(Q.Count==0) return null;
        if(Q.Peek()=="x"){
            Q.Dequeue();
            return null;
        }


        TreeNode node = new TreeNode(Int32.Parse(Q.Peek()));
        Q.Dequeue();

        node.left = tree(ref Q);
        node.right = tree(ref Q);

        return node;
    }
}
